Compile error missing bsd/sys/time.h include

Hi all,

I'm getting the following error when attempting to compile (076c837) for my platform:
```
../../src/ba-transport.c:27:10: fatal error: bsd/sys/time.h: No such file or directory
   27 | #include <bsd/sys/time.h>
      |          ^~~~~~~~~~~~~~~~
```

My platform does not have libbsd available during configuration:
```
checking for libbsd >= 0.8... no
```

Removing line 27 from `ba-transport.c` fixes the issue.

No worries! :)

As an unrelated side note. I'm also getting these as warnings in a few places:
```
bluez.c:996:6: warning: ‘g_memdup’ is deprecated: Use 'g_memdup2' instead [-Wdeprecated-declarations]
  996 |      g_memdup(&addr, sizeof(addr)), seps = g_array_new(FALSE, FALSE, sizeof(sep)));
      |      ^~~~~~~~
```
```
bluez.c:147:2: warning: ‘g_memdup’ is deprecated: Use 'g_memdup2' instead [-Wdeprecated-declarations]
  147 |  capabilities = g_memdup(data, size);
      |  ^~~~~~~~~~~~
```

--- a/src/a2dp-aac.c
+++ b/src/a2dp-aac.c
@@ -42,9 +42,9 @@ void a2dp_aac_transport_set_codec(struct ba_transport *t) {
 
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S16_2LE;
 	t->a2dp.pcm.channels = a2dp_codec_lookup_channels(codec,
-			((a2dp_aac_t *)t->a2dp.configuration)->channels, false);
+			t->a2dp.configuration.aac.channels, false);
 	t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-			AAC_GET_FREQUENCY(*(a2dp_aac_t *)t->a2dp.configuration), false);
+			AAC_GET_FREQUENCY(t->a2dp.configuration.aac), false);
 
 }
 
@@ -60,7 +60,7 @@ static void *a2dp_aac_enc_thread(struct ba_transport_thread *th) {
 	AACENC_InfoStruct aacinf;
 	AACENC_ERROR err;
 
-	const a2dp_aac_t *configuration = (a2dp_aac_t *)t->a2dp.configuration;
+	const a2dp_aac_t *configuration = &t->a2dp.configuration.aac;
 	const unsigned int bitrate = AAC_GET_BITRATE(*configuration);
 	const unsigned int channels = t->a2dp.pcm.channels;
 	const unsigned int samplerate = t->a2dp.pcm.sampling;

--- a/src/a2dp-aptx-hd.c
+++ b/src/a2dp-aptx-hd.c
@@ -38,9 +38,9 @@ void a2dp_aptx_hd_transport_set_codec(struct ba_transport *t) {
 
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S24_4LE;
 	t->a2dp.pcm.channels = a2dp_codec_lookup_channels(codec,
-			((a2dp_aptx_hd_t *)t->a2dp.configuration)->aptx.channel_mode, false);
+			t->a2dp.configuration.aptx_hd.aptx.channel_mode, false);
 	t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-			((a2dp_aptx_hd_t *)t->a2dp.configuration)->aptx.frequency, false);
+			t->a2dp.configuration.aptx_hd.aptx.frequency, false);
 
 }
 

--- a/src/a2dp-aptx.c
+++ b/src/a2dp-aptx.c
@@ -36,9 +36,9 @@ void a2dp_aptx_transport_set_codec(struct ba_transport *t) {
 
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S16_2LE;
 	t->a2dp.pcm.channels = a2dp_codec_lookup_channels(codec,
-			((a2dp_aptx_t *)t->a2dp.configuration)->channel_mode, false);
+			t->a2dp.configuration.aptx.channel_mode, false);
 	t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-			((a2dp_aptx_t *)t->a2dp.configuration)->frequency, false);
+			t->a2dp.configuration.aptx.frequency, false);
 
 }
 

--- a/src/a2dp-codecs.h
+++ b/src/a2dp-codecs.h
@@ -350,18 +350,6 @@ typedef struct {
 		.codec_id1 = (((c) >> 8) & 0xff), \
 	}
 
-typedef struct {
-	uint8_t reserved;
-	uint8_t target_codec_level2;
-	uint8_t target_codec_level1;
-	uint8_t initial_codec_level2;
-	uint8_t initial_codec_level1;
-	uint8_t sra_max_rate;
-	uint8_t sra_avg_time;
-	uint8_t good_working_level2;
-	uint8_t good_working_level1;
-} __attribute__ ((packed)) a2dp_aptx_ll_new_caps_t;
-
 #if defined(__BYTE_ORDER) && defined(__LITTLE_ENDIAN) && \
 	__BYTE_ORDER == __LITTLE_ENDIAN
 
@@ -432,7 +420,6 @@ typedef struct {
 	uint8_t bidirect_link:1;
 	uint8_t has_new_caps:1;
 	uint8_t reserved:6;
-	a2dp_aptx_ll_new_caps_t new_caps[0];
 } __attribute__ ((packed)) a2dp_aptx_ll_t;
 
 typedef struct {
@@ -518,7 +505,6 @@ typedef struct {
 	uint8_t reserved:6;
 	uint8_t has_new_caps:1;
 	uint8_t bidirect_link:1;
-	a2dp_aptx_ll_new_caps_t new_caps[0];
 } __attribute__ ((packed)) a2dp_aptx_ll_t;
 
 typedef struct {
@@ -537,5 +523,31 @@ typedef struct {
 #else
 # error "Unknown byte order"
 #endif
+
+typedef struct {
+	a2dp_aptx_ll_t aptx_ll;
+	uint8_t reserved;
+	uint8_t target_codec_level2;
+	uint8_t target_codec_level1;
+	uint8_t initial_codec_level2;
+	uint8_t initial_codec_level1;
+	uint8_t sra_max_rate;
+	uint8_t sra_avg_time;
+	uint8_t good_working_level2;
+	uint8_t good_working_level1;
+} __attribute__ ((packed)) a2dp_aptx_ll_new_t;
+
+typedef union {
+	a2dp_sbc_t sbc;
+	a2dp_mpeg_t mpeg;
+	a2dp_aac_t aac;
+	a2dp_atrac_t atrac;
+	a2dp_aptx_t aptx;
+	a2dp_faststream_t faststream;
+	a2dp_aptx_ll_t aptx_ll;
+	a2dp_aptx_ll_new_t aptx_ll_new;
+	a2dp_aptx_hd_t aptx_hd;
+	a2dp_ldac_t ldac;
+} a2dp_t;
 
 #endif

--- a/src/a2dp-faststream.c
+++ b/src/a2dp-faststream.c
@@ -39,16 +39,16 @@ void a2dp_faststream_transport_set_codec(struct ba_transport *t) {
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S16_2LE;
 	t->a2dp.pcm_bc.format = BA_TRANSPORT_PCM_FORMAT_S16_2LE;
 
-	if (((a2dp_faststream_t *)t->a2dp.configuration)->direction & FASTSTREAM_DIRECTION_MUSIC) {
+	if (t->a2dp.configuration.faststream.direction & FASTSTREAM_DIRECTION_MUSIC) {
 		t->a2dp.pcm.channels = 2;
 		t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-				((a2dp_faststream_t *)t->a2dp.configuration)->frequency_music, false);
+				t->a2dp.configuration.faststream.frequency_music, false);
 	}
 
-	if (((a2dp_faststream_t *)t->a2dp.configuration)->direction & FASTSTREAM_DIRECTION_VOICE) {
+	if (t->a2dp.configuration.faststream.direction & FASTSTREAM_DIRECTION_VOICE) {
 		t->a2dp.pcm_bc.channels = 1;
 		t->a2dp.pcm_bc.sampling = a2dp_codec_lookup_frequency(codec,
-				((a2dp_faststream_t *)t->a2dp.configuration)->frequency_voice, true);
+				t->a2dp.configuration.faststream.frequency_voice, true);
 	}
 
 }
@@ -66,8 +66,8 @@ static void *a2dp_faststream_enc_thread(struct ba_transport_thread *th) {
 	struct ba_transport_pcm *t_a2dp_pcm = is_voice ? &t->a2dp.pcm_bc : &t->a2dp.pcm;
 
 	sbc_t sbc;
-	if ((errno = -sbc_init_a2dp_faststream(&sbc, 0, t->a2dp.configuration,
-					t->a2dp.codec->capabilities_size, is_voice)) != 0) {
+	if ((errno = -sbc_init_a2dp_faststream(&sbc, 0, &t->a2dp.configuration.faststream,
+					sizeof(t->a2dp.configuration.faststream), is_voice)) != 0) {
 		error("Couldn't initialize FastStream SBC codec: %s", strerror(errno));
 		goto fail_init;
 	}
@@ -186,8 +186,8 @@ static void *a2dp_faststream_dec_thread(struct ba_transport_thread *th) {
 	struct ba_transport_pcm *t_a2dp_pcm = is_voice ? &t->a2dp.pcm_bc : &t->a2dp.pcm;
 
 	sbc_t sbc;
-	if ((errno = -sbc_init_a2dp_faststream(&sbc, 0, t->a2dp.configuration,
-					t->a2dp.codec->capabilities_size, is_voice)) != 0) {
+	if ((errno = -sbc_init_a2dp_faststream(&sbc, 0, &t->a2dp.configuration.faststream,
+					sizeof(t->a2dp.configuration.faststream), is_voice)) != 0) {
 		error("Couldn't initialize FastStream SBC codec: %s", strerror(errno));
 		goto fail_init;
 	}
@@ -267,17 +267,17 @@ int a2dp_faststream_transport_start(struct ba_transport *t) {
 	int rv = 0;
 
 	if (t->type.profile & BA_TRANSPORT_PROFILE_A2DP_SOURCE) {
-		if (((a2dp_faststream_t *)t->a2dp.configuration)->direction & FASTSTREAM_DIRECTION_MUSIC)
+		if (t->a2dp.configuration.faststream.direction & FASTSTREAM_DIRECTION_MUSIC)
 			rv |= ba_transport_thread_create(th_enc, a2dp_faststream_enc_thread, "ba-a2dp-fs-m", true);
-		if (((a2dp_faststream_t *)t->a2dp.configuration)->direction & FASTSTREAM_DIRECTION_VOICE)
+		if (t->a2dp.configuration.faststream.direction & FASTSTREAM_DIRECTION_VOICE)
 			rv |= ba_transport_thread_create(th_dec, a2dp_faststream_dec_thread, "ba-a2dp-fs-v", false);
 		return rv;
 	}
 
 	if (t->type.profile & BA_TRANSPORT_PROFILE_A2DP_SINK) {
-		if (((a2dp_faststream_t *)t->a2dp.configuration)->direction & FASTSTREAM_DIRECTION_MUSIC)
+		if (t->a2dp.configuration.faststream.direction & FASTSTREAM_DIRECTION_MUSIC)
 			rv |= ba_transport_thread_create(th_dec, a2dp_faststream_dec_thread, "ba-a2dp-fs-m", true);
-		if (((a2dp_faststream_t *)t->a2dp.configuration)->direction & FASTSTREAM_DIRECTION_VOICE)
+		if (t->a2dp.configuration.faststream.direction & FASTSTREAM_DIRECTION_VOICE)
 			rv |= ba_transport_thread_create(th_enc, a2dp_faststream_enc_thread, "ba-a2dp-fs-v", false);
 		return rv;
 	}

--- a/src/a2dp-ldac.c
+++ b/src/a2dp-ldac.c
@@ -44,9 +44,9 @@ void a2dp_ldac_transport_set_codec(struct ba_transport *t) {
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S32_4LE;
 
 	t->a2dp.pcm.channels = a2dp_codec_lookup_channels(codec,
-			((a2dp_ldac_t *)t->a2dp.configuration)->channel_mode, false);
+			t->a2dp.configuration.ldac.channel_mode, false);
 	t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-			((a2dp_ldac_t *)t->a2dp.configuration)->frequency, false);
+			t->a2dp.configuration.ldac.frequency, false);
 
 }
 
@@ -74,7 +74,7 @@ static void *a2dp_ldac_enc_thread(struct ba_transport_thread *th) {
 
 	pthread_cleanup_push(PTHREAD_CLEANUP(ldac_ABR_free_handle), handle_abr);
 
-	const a2dp_ldac_t *configuration = (a2dp_ldac_t *)t->a2dp.configuration;
+	const a2dp_ldac_t *configuration = &t->a2dp.configuration.ldac;
 	const size_t sample_size = BA_TRANSPORT_PCM_FORMAT_BYTES(t->a2dp.pcm.format);
 	const unsigned int channels = t->a2dp.pcm.channels;
 	const unsigned int samplerate = t->a2dp.pcm.sampling;
@@ -255,7 +255,7 @@ static void *a2dp_ldac_dec_thread(struct ba_transport_thread *th) {
 
 	pthread_cleanup_push(PTHREAD_CLEANUP(ldacBT_free_handle), handle);
 
-	const a2dp_ldac_t *configuration = (a2dp_ldac_t *)t->a2dp.configuration;
+	const a2dp_ldac_t *configuration = &t->a2dp.configuration.ldac;
 	const size_t sample_size = BA_TRANSPORT_PCM_FORMAT_BYTES(t->a2dp.pcm.format);
 	const unsigned int channels = t->a2dp.pcm.channels;
 	const unsigned int samplerate = t->a2dp.pcm.sampling;

--- a/src/a2dp-mpeg.c
+++ b/src/a2dp-mpeg.c
@@ -45,9 +45,9 @@ void a2dp_mpeg_transport_set_codec(struct ba_transport *t) {
 
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S16_2LE;
 	t->a2dp.pcm.channels = a2dp_codec_lookup_channels(codec,
-			((a2dp_mpeg_t *)t->a2dp.configuration)->channel_mode, false);
+			t->a2dp.configuration.mpeg.channel_mode, false);
 	t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-			((a2dp_mpeg_t *)t->a2dp.configuration)->frequency, false);
+			t->a2dp.configuration.mpeg.frequency, false);
 
 }
 
@@ -68,7 +68,7 @@ static void *a2dp_mp3_enc_thread(struct ba_transport_thread *th) {
 
 	pthread_cleanup_push(PTHREAD_CLEANUP(lame_close), handle);
 
-	const a2dp_mpeg_t *configuration = (a2dp_mpeg_t *)t->a2dp.configuration;
+	const a2dp_mpeg_t *configuration = &t->a2dp.configuration.mpeg;
 	const unsigned int channels = t->a2dp.pcm.channels;
 	const unsigned int samplerate = t->a2dp.pcm.sampling;
 	MPEG_mode mode = NOT_SET;
@@ -461,7 +461,7 @@ int a2dp_mpeg_transport_start(struct ba_transport *t) {
 
 	if (t->type.profile & BA_TRANSPORT_PROFILE_A2DP_SOURCE) {
 #if ENABLE_MP3LAME
-		if (((a2dp_mpeg_t *)t->a2dp.configuration)->layer == MPEG_LAYER_MP3)
+		if (t->a2dp.configuration.mpeg.layer == MPEG_LAYER_MP3)
 			return ba_transport_thread_create(&t->thread_enc, a2dp_mp3_enc_thread, "ba-a2dp-mp3", true);
 #endif
 	}
@@ -470,7 +470,7 @@ int a2dp_mpeg_transport_start(struct ba_transport *t) {
 #if ENABLE_MPG123
 		return ba_transport_thread_create(&t->thread_dec, a2dp_mpeg_dec_thread, "ba-a2dp-mpeg", true);
 #elif ENABLE_MP3LAME
-		if (((a2dp_mpeg_t *)t->a2dp.configuration)->layer == MPEG_LAYER_MP3)
+		if (t->a2dp.configuration.mpeg.layer == MPEG_LAYER_MP3)
 			return ba_transport_thread_create(&t->thread_dec, a2dp_mpeg_dec_thread, "ba-a2dp-mp3", true);
 #endif
 	}

--- a/src/a2dp-sbc.c
+++ b/src/a2dp-sbc.c
@@ -41,9 +41,9 @@ void a2dp_sbc_transport_set_codec(struct ba_transport *t) {
 
 	t->a2dp.pcm.format = BA_TRANSPORT_PCM_FORMAT_S16_2LE;
 	t->a2dp.pcm.channels = a2dp_codec_lookup_channels(codec,
-			((a2dp_sbc_t *)t->a2dp.configuration)->channel_mode, false);
+			t->a2dp.configuration.sbc.channel_mode, false);
 	t->a2dp.pcm.sampling = a2dp_codec_lookup_frequency(codec,
-			((a2dp_sbc_t *)t->a2dp.configuration)->frequency, false);
+			t->a2dp.configuration.sbc.frequency, false);
 
 }
 
@@ -56,8 +56,8 @@ static void *a2dp_sbc_enc_thread(struct ba_transport_thread *th) {
 	struct io_poll io = { .timeout = -1 };
 
 	sbc_t sbc;
-	if ((errno = -sbc_init_a2dp(&sbc, 0, t->a2dp.configuration,
-					t->a2dp.codec->capabilities_size)) != 0) {
+	if ((errno = -sbc_init_a2dp(&sbc, 0, &t->a2dp.configuration.sbc,
+					sizeof(t->a2dp.configuration.sbc))) != 0) {
 		error("Couldn't initialize SBC codec: %s", strerror(errno));
 		goto fail_init;
 	}
@@ -68,7 +68,7 @@ static void *a2dp_sbc_enc_thread(struct ba_transport_thread *th) {
 	pthread_cleanup_push(PTHREAD_CLEANUP(ffb_free), &pcm);
 	pthread_cleanup_push(PTHREAD_CLEANUP(sbc_finish), &sbc);
 
-	const a2dp_sbc_t *configuration = (a2dp_sbc_t *)t->a2dp.configuration;
+	const a2dp_sbc_t *configuration = &t->a2dp.configuration.sbc;
 	const size_t sbc_frame_samples = sbc_get_codesize(&sbc) / sizeof(int16_t);
 	const unsigned int channels = t->a2dp.pcm.channels;
 	const unsigned int samplerate = t->a2dp.pcm.sampling;
@@ -223,8 +223,8 @@ static void *a2dp_sbc_dec_thread(struct ba_transport_thread *th) {
 	};
 
 	sbc_t sbc;
-	if ((errno = -sbc_init_a2dp(&sbc, 0, t->a2dp.configuration,
-					t->a2dp.codec->capabilities_size)) != 0) {
+	if ((errno = -sbc_init_a2dp(&sbc, 0, &t->a2dp.configuration.sbc,
+					sizeof(t->a2dp.configuration.sbc))) != 0) {
 		error("Couldn't initialize SBC codec: %s", strerror(errno));
 		goto fail_init;
 	}

--- a/src/ba-transport.c
+++ b/src/ba-transport.c
@@ -24,8 +24,6 @@
 #include <bluetooth/hci.h>
 #include <bluetooth/hci_lib.h>
 
-#include <bsd/sys/time.h>
-
 #include <gio/gio.h>
 #include <gio/gunixfdlist.h>
 #include <glib-object.h>
@@ -655,7 +653,7 @@ struct ba_transport *ba_transport_new_a2dp(
 	t->type = type;
 
 	t->a2dp.codec = codec;
-	t->a2dp.configuration = g_memdup(configuration, codec->capabilities_size);
+	memcpy(&t->a2dp.configuration, configuration, codec->capabilities_size);
 	t->a2dp.state = BLUEZ_A2DP_TRANSPORT_STATE_IDLE;
 
 	transport_pcm_init(&t->a2dp.pcm,
@@ -885,7 +883,6 @@ void ba_transport_unref(struct ba_transport *t) {
 	if (t->type.profile & BA_TRANSPORT_PROFILE_MASK_A2DP) {
 		transport_pcm_free(&t->a2dp.pcm);
 		transport_pcm_free(&t->a2dp.pcm_bc);
-		free(t->a2dp.configuration);
 	}
 	else if (t->type.profile & BA_TRANSPORT_PROFILE_MASK_SCO) {
 		if (t->sco.rfcomm != NULL)
@@ -965,7 +962,7 @@ int ba_transport_select_codec_a2dp(
 
 	/* the same codec with the same configuration already selected */
 	if (t->type.codec == sep->codec_id &&
-			memcmp(sep->configuration, t->a2dp.configuration, sep->capabilities_size) == 0)
+			memcmp(sep->configuration, &t->a2dp.configuration, sep->capabilities_size) == 0)
 		goto final;
 
 	GError *err = NULL;

--- a/src/ba-transport.h
+++ b/src/ba-transport.h
@@ -22,6 +22,7 @@
 #include <stdint.h>
 
 #include "a2dp.h"
+#include "a2dp-codecs.h"
 #include "ba-device.h"
 #include "ba-rfcomm.h"
 #include "bluez.h"
@@ -262,7 +263,7 @@ struct ba_transport {
 			/* audio codec configuration capabilities */
 			const struct a2dp_codec *codec;
 			/* selected audio codec configuration */
-			uint8_t *configuration;
+			a2dp_t configuration;
 
 			/* delay reported by the AVDTP */
 			uint16_t delay;

--- a/src/ofono.c
+++ b/src/ofono.c
@@ -209,6 +209,7 @@ static void ofono_card_add(const char *dbus_sender, const char *card,
 	struct ba_device *d = NULL;
 	struct ba_transport *t = NULL;
 
+	struct ofono_card_data *ocd = NULL;
 	const char *key = NULL;
 	GVariant *value = NULL;
 	bdaddr_t addr_dev = { 0 };
@@ -255,19 +256,22 @@ static void ofono_card_add(const char *dbus_sender, const char *card,
 		goto fail;
 	}
 
-	struct ofono_card_data ocd = {
-		.hci_dev_id = hci_dev_id,
-		.bt_addr = addr_dev,
-	};
+	if ((ocd = malloc(sizeof(*ocd))) == NULL) {
+		error("Couldn't create oFono card data: %s", strerror(errno));
+		goto fail;
+	}
 
-	snprintf(ocd.transport_path, sizeof(ocd.transport_path), "/ofono%s", card);
-	if ((t = ofono_transport_new(d, ttype, dbus_sender, ocd.transport_path)) == NULL) {
+	ocd->hci_dev_id = hci_dev_id;
+	ocd->bt_addr = addr_dev;
+	snprintf(ocd->transport_path, sizeof(ocd->transport_path), "/ofono%s", card);
+
+	if ((t = ofono_transport_new(d, ttype, dbus_sender, ocd->transport_path)) == NULL) {
 		error("Couldn't create new transport: %s", strerror(errno));
 		goto fail;
 	}
 
-	g_hash_table_insert(ofono_card_data_map, g_strdup(card),
-			g_memdup(&ocd, sizeof(ocd)));
+	g_hash_table_insert(ofono_card_data_map, g_strdup(card), ocd);
+	ocd = NULL;
 
 	ba_transport_start(t);
 
@@ -280,6 +284,7 @@ fail:
 		ba_transport_unref(t);
 	if (value != NULL)
 		g_variant_unref(value);
+	free(ocd);
 }
 
 /**

--- a/utils/a2dpconf.c
+++ b/utils/a2dpconf.c
@@ -316,12 +316,12 @@ static void dump_aptx_hd(const void *blob, size_t size) {
 static void dump_aptx_ll(const void *blob, size_t size) {
 
 	const a2dp_aptx_ll_t *aptx_ll = blob;
-	const a2dp_aptx_ll_new_caps_t *aptx_ll_new_caps = (void *)(aptx_ll + 1);
+	const a2dp_aptx_ll_new_t *aptx_ll_new_caps = blob;
 
-	size_t new_caps_size = 0;
+	size_t conf_size = sizeof(*aptx_ll);
 	if (size >= sizeof(*aptx_ll) && aptx_ll->has_new_caps)
-		new_caps_size = sizeof(*aptx_ll_new_caps);
-	if (check_blob_size(sizeof(*aptx_ll) + new_caps_size, size) == -1)
+		conf_size = sizeof(*aptx_ll_new_caps);
+	if (check_blob_size(conf_size, size) == -1)
 		return;
 
 	printf("aptX LL (Sprint) <hex:%s> {\n", bintohex(blob, size));
@@ -334,7 +334,7 @@ static void dump_aptx_ll(const void *blob, size_t size) {
 			aptx_ll->has_new_caps ? "true" : "false",
 			aptx_ll->bidirect_link ? "true" : "false");
 
-	if (new_caps_size > 0)
+	if (aptx_ll->has_new_caps)
 		printf(""
 				"  <reserved>:8\n"
 				"  target-codec-level:16 = %u\n"
@@ -407,8 +407,7 @@ static struct {
 	{ A2DP_CODEC_VENDOR_APTX_AD, -1, dump_vendor },
 	{ A2DP_CODEC_VENDOR_APTX_HD, sizeof(a2dp_aptx_hd_t), dump_aptx_hd },
 	{ A2DP_CODEC_VENDOR_APTX_LL, sizeof(a2dp_aptx_ll_t), dump_aptx_ll },
-	{ A2DP_CODEC_VENDOR_APTX_LL,
-		sizeof(a2dp_aptx_ll_t) + sizeof(a2dp_aptx_ll_new_caps_t), dump_aptx_ll },
+	{ A2DP_CODEC_VENDOR_APTX_LL, sizeof(a2dp_aptx_ll_new_t), dump_aptx_ll },
 	{ A2DP_CODEC_VENDOR_FASTSTREAM, sizeof(a2dp_faststream_t), dump_faststream },
 	{ A2DP_CODEC_VENDOR_LDAC, sizeof(a2dp_ldac_t), dump_ldac },
 	{ A2DP_CODEC_VENDOR_LHDC, -1, dump_vendor },

